### PR TITLE
ADDED: /models and /model Commands to List and Switch Models During Live Chat Session

### PR DESCRIPTION
**Pull request by fire17**
Hi there, I was missing the option to change the model during a run,
for example when i know i want to ask a simple question, that is somewhat related (needs context)
but doesnt need gpt4, + It's also better if you want to use `--model gpt-4-1106-preview` without exiting aider and starting over

So here it is **ADDED** `/models` , `/model {alias}` and `/m {alias}` to change see and change models:
image

This was accomplished using aider to help me and it actually managed to preduce the changes, and was working after some back and forth with it!

* Do note: when using `gpt-4-1106-preview` i found that it sometimes doesnt obey the ```<source>python``` and instead uses ` ```python` and then aider asks if i want the changes to go into a new file called "```python"
   please see how this can be fixed @paul-gauthier 

For more info check out my original issue requesting this: #402

Hope you like it :)
Thanks a lot and all the best 💛

### CODE REVIEW

File: aider/models/model.py [hunk 1]
```diff
@@ -17,6 +17,7 @@ class Model:
 
     @classmethod
     def create(cls, name, client=None):
+        # print(f"Creating model with name: {name}")  # Debug print statement
```
Comment (fire17) [inline aider/models/model.py:20]: Can be ignored

**Comment by fire17** [comment]
@paul-gauthier I've just pulled and merged everything from your main branch (now with unified diffs)
I checked and it's working great, so this PR up-to-date and ready :)
Please see the "Files Changed" section, I made sure this PR is slim
All the best and Happy Holydays!

**Comment by joshuavial** [comment]
Nice, I really like this feature and have given the branch a test locally. Some thoughts in no real order

- love the /m command
- when I switch a model I see the 'switched to model: ..' message twice
- not working with openrouter models so I made a PR with some fixes - could do with a little more work to make alises work
- I would love it if we could add the supported models to auto complete, would make switching openrouter models much easier
- I think the model files should contain logic about which coders work well with which models and either warn the user or switch to the default coder for a model when switched - if you're using udiff and switch to -3 you probably want to bounce over to whole - so maybe a /coder /coders /c set of commands would be desirable?
- Haven't throught through what should happen to chat history when switching models / coders - it's probably fine but I'd want to test it out
- would love to see some unit tests covering the commands

**Comment by fire17** [comment]
Thanks for the response! Happy you like it ❤️ @joshuavial 
Lemme try to address your thoughts (and add a couple of my own)

About openrouter, Im not familar with it, understand it opens up aider to more then just oai, hope you could handle this if possible. Is it just a nexus for more models, or does it let me turn a custom local llm into an api i can call appon?

Fixes and Improvements for `/models`:
1. I obviously noticed the double "switched to.." message, havn't figured how to fix it yet
    - it would be awesome if some tests could process/pipe aider's live outputs/logs. Then we can ask aider to fix this (loop with test runs) until it is solved (See more about PIPELINES)
2. I think `/m` and `/model` currently only accept alias, it would be best if no alias is found, use the model by name
3. Recommand adding a `/udiff` command (or have it be apart of `/coders` as you said)
    - If udiff  is required for some models, and should be disabled in others, it should go into models file, its already a dict, and it can have an optional prop for that.
4. Models can have a `.commands` :list prop that runs any `/cmd` in sequence - if all settings are available throught the /commands, then any model can be set to anything. ie `commands:["/udiff on",]` or `commands:["/udiff off",]` for relevant models
5. Adding Models to Auto Complete
     - Is there a nice pattern we can use to easily add any future feature of aider to autocomplete easily? for example adding a `@aider.autocomplete(cmds=["/m","/model"])` attribute above the models dictionary or something like that, that would not require hardcoded/manual inserts 
6. (Later) The models file, should be allowed to be edited by the user to custom, local or cloud, llm apis
     - Add Vision models support 
     - Fetch new models from openrouter and or official github:aider repo
     - Add support for PIPELINES. (Really strong useage of aider, will explain further in the following comment)
     
I'll try to tackle some of these, but some might need more help, design and consideration
Meanwhile i want to post about Pipelines.. Lemme know what you think about it, and about what i've summerized, and
If there's anything else that should be addressed.

Thanks a lot and all the best!

**Comment by fire17** [comment]
Addressing your notes led me to Pipelines, which i want to explore
Will probably start a new PR on this, but want to get your initial reaction and thoughts @joshuavial @paul-gauthier 

Would like us to lay the ground work for enabling PIPELINES.. what are Pipelines?

Aider is currently a dev companion, but it is **us** who check for and iterate with it. Imo it is the best general solution, but for specific (and some generic) cases we can do much better. 
The same way aider has a mechenism to check that the code changes were made correctly (and if not rerun to get correct code),
we can check for **solved conditions**, and have aider iterate until the conditions are solved (perhaps with a default /undo at the end if conditions were not satisfied). In practice this can be accomplished with: 1. build a pipeline by chaining functions, the `/run` cmd, testing the outputs, and using function calling (or current aider mechanism) to retrigger an iteration.

Using a pipeline should be as simple as choosing a `/model` `/pipeline issue-solver github.com/paul-gauthier/aider/401/#issuecomment-1871852269`
Here we use the `issue-solver` pipeline, with url to this PR, with issue comment to direct to a single message within a PR/Issue. Some more CI/CD examples:

Pipelines **EXAMPLES** : (much like github actions, can have multiple consecutive steps)
* `compiles: # Runs a simple test to see that the app runs without errors`
* `test-pass`: [switching_model_one_output, *production_tests] # Create & Choose Tests (like fixing the double "switching to.." msg). Iterate aider until all tests pass.
* `issue-solver`: magical dev exp, point aider to this issue (or pr thread) , and everything we mentioned should be turned into a list of aider missions, tests. have aider iterate until this issue is resolved, with code in new branch. include a file explaining the full progress report
* `researcher:` # Does online research from code guides, github repos, official docs, yt, etc, and adds a smart summerized research to the context before running + save to local skill lib
* `local-skill-lib:` find relevant skills [RAGish] from local skill lib and load to virtual context
* `global-skill-lib:` find relevant skills from the global community. Users contribute knoweledge from their skill local lib, and the oss communtiy marks packages as safe and rank them as useful **This might be end the days of scavanging github or online for solutions**
* `security:` checks and solves all security and vulnerability issues

I think anyone of these pipelines can be an absolute gamechanger in dev history
please tell me if there are any specific pipelines you like, and if you can think of there ways to use Aider in a more advanced or powerfull way

I hope that in early 2024, we would all have access to a UI like MakeReal + Admin Dash, that connects with Aider under the hood, uses the most advanced pipelines, delivers advance products, with the highest quality, least cost, and minimal time. We will start designing from top to bottom, and have proper tools to visualize, and rapidly modify any <> of our code, from mission to design to the low level code and everything inbetween

```
# aider-pipelines-v0.1 
pipelines = { "issue-solver":
            {
                "kwargs": {"target": "<url or alias for pr, issue or comment>", ** },
                "content": fetchContent( kwargs["target"] ),
                "id": makeIssueId( content, uuid() ),
                "branch": {"create_new":True, "name":f"issue_solver_{id}"},
                "init": [ createMissions( id, content), updateTests( id ), ], 
                "pre_run": [f"/add missions_{id}.txt missions_{id}_test*", ],
                "model" {"model":"gpt-4-1106-preview", "cmds":["/udiff true"], max_cost:2.0, undo_if_failed:False},
                "pass": [ f"missions_{id}_test*" ] # This pipeline will run until all tests have passed or cost exeedes 2 dollars
            }
            , # other pipelines
        }
```

**Comment by joshuavial** [comment]
@fire17 I've pushed a bunch more changes to my model branch including
- merge in current main
- moving the model_tables data into models/openai
- remove the duplicate io output when switching model
- fixing token count when manually using gpt-4vision model
- creating a new coder when switching models (so `/m 3` will switch to whole coder, `/m 4` will switch to udiff etc.)

Did you want me to make a separate pull request into the main aider repo with all that work? The other option would be for you to merge it into your main branch which should get this PR up to date.

I don't have many thoughts about pipelines at this stage - might be good to start a separate issue and I'm happy to review it when I have a bit more headspace.

**Comment by paul-gauthier** [comment]
@fire17 and @joshuavial, where do we stand on this PR? It sounds like @joshuavial has a related change in his repo. Which one should we be looking to merge?

**Comment by joshuavial** [comment]
was thinking about this yesterday, I'm really enjoying the switching models feature and have done a bunch of changes in my branch. I'm fine if @fire17 wants to merge it in here, otherwise I can start a separate PR which includes all of this work. Last thing I want to add is autocomplete for the openrouter models as they are a pain to type.

**Comment by paul-gauthier** [comment]
Closing since #468 seems to replace this.

**Comment by fire17** [comment]
@joshuavial that's why I've added the aliasing
But autocomplete for alias/fullname would be nice

@paul-gauthier I honestly don't mind who does the contribution, as long as it works well with the up-to-date Aider

Hope my PR gave inspiration, and hope to see everything included soon. Will check out #486
And hope that it has everything I needed. Fast Switching /m, list all /models and very important to update the /tokens calculation for correct maxContext and price-per-token

**Pull request closed without merging**